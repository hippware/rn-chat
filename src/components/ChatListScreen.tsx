import React from 'react'
import {View, StyleSheet, FlatList, Image} from 'react-native'
import {observer, inject} from 'mobx-react/native'
import {Actions} from 'react-native-router-flux'
import Screen from './Screen'
import MessageButton from './MessageButton'
import ChatCard from './ChatCard'
import ListFooter from './ListFooter'
import {RText} from './common'
import {colors} from '../constants'
import {IWocky} from 'wocky-client'

const footerImage = require('../../images/graphicEndMsgs.png')

type Props = {
  wocky?: IWocky
}

@inject('wocky')
@observer
class ChatListScreen extends React.Component<Props> {
  list: any

  scrollTo = params => {
    this.list.scrollTo(params)
  }

  componentDidMount() {
    this.props.wocky!.loadChats()
  }

  renderItem = ({item}) => <ChatCard item={item} onPress={i => Actions.chat({item: i.id})} />

  keyExtractor = item => `${item.id}`

  render() {
    const {list: chats, unread: num} = this.props.wocky!.chats
    return (
      <Screen>
        <FlatList
          style={{flex: 1, flexDirection: 'column'}}
          ref={l => (this.list = l)}
          contentContainerStyle={{marginTop: num ? 47 : 10}}
          data={chats}
          initialNumToRender={6}
          ListFooterComponent={
            chats.length ? <ListFooter footerImage={footerImage} finished /> : null
          }
          ListEmptyComponent={EmptyComponent}
          renderItem={this.renderItem}
          keyExtractor={this.keyExtractor}
        />
        <MessageButton />
        {!!num && (
          <View style={styles.button}>
            <RText weight="Italic" color="white">
              New Messages
            </RText>
          </View>
        )}
      </Screen>
    )
  }
}

export default ChatListScreen

const EmptyComponent = () => (
  <View
    style={{
      flex: 1,
      alignItems: 'center',
      justifyContent: 'center',
      marginTop: 150,
      backgroundColor: 'transparent',
    }}
  >
    <RText size={18} color={colors.PINKISH_GREY} style={{textAlign: 'center'}}>
      {'No messages yet\r\nWhy not start a conversation?'}
    </RText>
    <Image source={require('../../images/botGray.png')} style={{marginTop: 20}} />
  </View>
)

const styles = StyleSheet.create({
  button: {
    position: 'absolute',
    right: 0,
    left: 0,
    top: 0,
    alignItems: 'center',
    justifyContent: 'center',
    height: 47,
    backgroundColor: 'rgba(254,92,108, 0.9)',
  },
})
