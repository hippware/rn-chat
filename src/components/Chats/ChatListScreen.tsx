import React from 'react'
import {View, Image, TouchableOpacity} from 'react-native'
import {observer, inject} from 'mobx-react/native'
import {Actions} from 'react-native-router-flux'
import ChatCard from './ChatCard'
import {RText} from '../common'
import {colors} from '../../constants'
import {IWocky} from 'wocky-client'
import DraggablePopupList from '../common/DraggablePopupList'

const sendMessageImg = require('../../../images/sendMessage.png')

type Props = {
  wocky?: IWocky
  isActive: boolean
}

@inject('wocky')
@observer
class ChatListScreen extends React.Component<Props> {
  static navigationOptions = {
    fadeNavConfig: {
      back: true,
      title: (
        <RText size={16} color={colors.DARK_PURPLE}>
          Messages
        </RText>
      ),
    },
  }

  // todo: convert to functional component with hooks, but mobx-react 6+ required
  componentDidMount() {
    this.props.wocky!.chats.loadChats()
  }

  renderItem = ({item}) => <ChatCard chat={item} onPress={i => Actions.chat({item: i.id})} />

  keyExtractor = item => `${item.id}`

  render() {
    const {wocky, isActive} = this.props
    const {chats} = wocky!
    return (
      <DraggablePopupList
        contentContainerStyle={{marginTop: chats.unreadCount > 0 ? 47 : 10}}
        data={chats.list.slice()}
        initialNumToRender={6}
        headerInner={
          <>
            <TouchableOpacity
              onPress={() => null}
              style={{
                borderColor: colors.PINK,
                borderWidth: 1,
                borderRadius: 12,
                flexDirection: 'row',
                padding: 10,
                alignItems: 'center',
                justifyContent: 'center',
                // todo: how do I do this with flex?
                width: 200,
                alignSelf: 'center',
              }}
            >
              <Image source={sendMessageImg} style={{marginRight: 8}} />
              <RText color={colors.PINK} size={14} style={{}}>
                Send Message
              </RText>
            </TouchableOpacity>
            <RText
              weight="Medium"
              size={16}
              style={{paddingLeft: 10, paddingTop: 25, paddingBottom: 10}}
            >
              Messages
            </RText>
          </>
        }
        // todo: figure out a flexible height setting
        ListEmptyComponent={
          <View style={{height: 200, alignItems: 'center', justifyContent: 'center'}}>
            <RText color={colors.DARK_GREY} size={16}>
              No messages
            </RText>
          </View>
        }
        renderItem={({item}) => (
          <ChatCard
            chat={item}
            onPress={() => Actions.chat({item: item.id})}
            style={{paddingHorizontal: 30}}
          />
        )}
        keyExtractor={({id}) => id}
        isActive={isActive}
      />
    )
  }
}

export default ChatListScreen
