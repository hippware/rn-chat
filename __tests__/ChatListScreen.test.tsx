import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import ChatListScreen from '../src/components/ChatListScreen'
import mockStore from './utils/mockStore'

describe('ChatListScreen', () => {
  it('renders with no data', () => {
    const tree = renderer.create(<ChatListScreen {...mockStore} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
