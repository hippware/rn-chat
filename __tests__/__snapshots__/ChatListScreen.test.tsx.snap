// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatListScreen renders with no data 1`] = `
<View
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <BVLinearGradient
    colors={
      Array [
        "rgb(255,255,255)",
        "rgb(241,242,244)",
        "rgb(243,244,246)",
      ]
    }
    locations={
      Array [
        0,
        0.2,
        1,
      ]
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      ListFooterComponent={null}
      contentContainerStyle={
        Object {
          "marginTop": 10,
        }
      }
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={6}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "flex": 1,
                "justifyContent": "center",
                "marginTop": 150,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "rgb(196,196,196)",
                    "fontFamily": "Roboto-Regular",
                    "fontSize": 18,
                  },
                  Object {
                    "textAlign": "center",
                  },
                ]
              }
            >
              No messages yet
Why not start a conversation?
            </Text>
            <Image
              source={
                Object {
                  "testUri": "../../../images/botGray.png",
                }
              }
              style={
                Object {
                  "marginTop": 20,
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(148,94,135)",
          "borderRadius": 27,
          "bottom": 40,
          "height": 54,
          "opacity": 1,
          "position": "absolute",
          "right": 40,
          "width": 54,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../images/iconNewMsg.png",
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
